python/mxnet/optimizer.py

```diff
@@ -670,9 +670,10 @@ def __init__(self, optimizer):
 
     def __call__(self, index, grad, weight):
         """Update weight given gradient and index"""
-        if index not in self.states:
-            self.states[index] = self.optimizer.create_state(index, weight)
-        self.optimizer.update(index, weight, grad, self.states[index])
+        key = (index, str(weight.context))
```

Comment:
This breaks optimizer state loading...

Are you sure we are not creating separate updater for different gpus already?